src/main/scala/scala/tools/refactoring/transformation/TreeTransformations.scala

```diff
@@ -177,40 +177,80 @@ trait TreeTransformations extends Transformations with TreeFactory {
     case t: global.Tree => t.pos = global.NoPosition; t
   }
 
-  def addImportTransformation(importsToAdd: Iterable[String]): Transformation[Tree, Tree] = {
-
-    import global._
-
-    def importTrees = {
-      val SplitAtDot = "(.*)\\.(.*?)".r
-      importsToAdd.map {
-        case SplitAtDot(pkg, name) => mkImportFromStrings(pkg, name)
-      }.toList
+  def addImportTransformation(importsToAdd: Seq[String]): Transformation[Tree, TextChange] = {
+    def splitImports(p: PackageDef, stats: List[Tree]) = {
+      val (imports, others) = stats partition (_.isInstanceOf[Import])
+      (p, imports map (_.asInstanceOf[Import]), others)
     }
 
-    val addImportStatement = once(findBestPackageForImports &> transformation[(PackageDef, List[Import], List[Tree]), Tree] {
+    def importsAsSrc(indent: String) = {
+      def escaped(imp: String) = imp.split('.').map(escapeScalaKeywordsForImport).mkString(".")
+      importsToAdd.map(imp ⇒ s"${indent}import ${escaped(imp)}").mkString("\n")
+    }
 
-      // For an empty PackageDef, with no exiting imports but with an Impl that has an annotation, we need to
-      // modify positions so that the annotation, which is not in the AST, doesn't get assigned to the PackageDef
-      // but to the Impl
-      case (p @ PackageDef(Ident(nme.EMPTY_PACKAGE_NAME), _), Nil, others @ (impl :: _)) if !impl.symbol.annotations.isEmpty =>
+    def insertAfter(pos: Position) = {
+      val lineNumber = pos.source.offsetToLine(pos.start)
+      val line = pos.source.lineToString(lineNumber)
+      val indent = line.takeWhile(Character.isWhitespace)
+      val insertPos = pos.source.lineToOffset(lineNumber) + line.length
+      TextChange(pos.source, insertPos, insertPos, "\n" + importsAsSrc(indent))
+    }
 
-        // The `pid` is invisible anyway, but by erasing its position we make sure
-        // it doesn't get any layout associated
-        val ignoredPid = p.pid setPos NoPosition
+    def insertBefore(pos: Position) = {
+      val lineNumber = pos.source.offsetToLine(pos.start)
+      val line = pos.source.lineToString(lineNumber)
+      val indent = line.takeWhile(Character.isWhitespace)
+      val insertPos = pos.source.lineToOffset(lineNumber)
+      TextChange(pos.source, insertPos, insertPos, importsAsSrc(indent) + "\n")
+    }
 
```

Comment:
just to clarify. `pos: Position` is a position in original file. So when I have multiple changes applicable at the same position their are applied in order in list. (I cannot find it but I think that there was a piece which reverts the list of changes before application which has a sense)